src/bt-pan-io.c

```diff
@@ -0,0 +1,345 @@
+                         17, /* interface name */
+                         0, NULL);
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_PAN, OPCODE_CONTROL_STATE_NTF);
+  if ((append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)state) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)error) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)local_role) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "0", ifname) < 0))
+    goto cleanup;
+
+  //padding
+  void* padding;
+  int len;
+  len = 17 - strnlen(ifname, 17);
+  if(len > 0){
+    padding = malloc(len);
+    if(!padding){
+      ALOGE_ERRNO("malloc");
+      goto cleanup;
+    }
+    if(append_to_pdu(&wbuf->buf.pdu, "M", padding,len) < 0)
+      goto cleanup;
+  }
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+static void
+connection_state_cb(btpan_connection_state_t state, bt_status_t error,
+                    const bt_bdaddr_t *bd_addr, int local_role, int remote_role)
+{
+  struct pdu_wbuf* wbuf;
+
+  wbuf = create_pdu_wbuf(1 + /* connection state */
+                         1 + /* status */
+                         6 + /* address */
+                         1 + /* local role */
+                         1, /* remote role */
+                         0, NULL);
+  if (!wbuf)
+    return;
+
+  init_pdu(&wbuf->buf.pdu, SERVICE_BT_PAN, OPCODE_CONNECTION_STATE_NTF);
+  if ((append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)state) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)error) < 0) ||
+      (append_bt_bdaddr_t(&wbuf->buf.pdu, bd_addr) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)local_role) < 0) ||
+      (append_to_pdu(&wbuf->buf.pdu, "C", (uint8_t)remote_role) < 0))
+    goto cleanup;
+
+  if (run_task(send_ntf_pdu, wbuf) < 0)
+    goto cleanup;
+
+  return;
+cleanup:
+  cleanup_pdu_wbuf(wbuf);
+}
+
+/*
+ * Commands/Responses
+ */
+
+static bt_status_t
+opcode_enable(const struct pdu* cmd)
+{
+  long off;
+  uint8_t local_role;
+  struct pdu_wbuf* wbuf;
+  bt_status_t status;
+
+  assert(btpan_interface);
+  assert(btpan_interface->enable);
+
+  off = read_pdu_at(cmd, 0, "C", &local_role);
+  if (off < 0)
+    return BT_STATUS_PARM_INVALID;
+
+  wbuf = create_pdu_wbuf(0, 0, NULL);
+  if (!wbuf)
+    return BT_STATUS_NOMEM;
+
+  status = btpan_interface->enable(local_role);
+  if (status != BT_STATUS_SUCCESS)
+    goto err_btpan_interface_enable;
+
+  init_pdu(&wbuf->buf.pdu, cmd->service, cmd->opcode);
+  send_pdu(wbuf);
+
+  return BT_STATUS_SUCCESS;
+err_btpan_interface_enable:
+  cleanup_pdu_wbuf(wbuf);
+  return status;
+}
+
+static bt_status_t
+opcode_get_local_role(const struct pdu* cmd)
+{
+  struct pdu_wbuf* wbuf;
+  uint8_t status;
+
+  assert(btpan_interface);
+  assert(btpan_interface->get_local_role);
+
+  wbuf = create_pdu_wbuf(1 /* local role */, 0, NULL);
+  if (!wbuf)
+    return BT_STATUS_NOMEM;
+
+  status = btpan_interface->get_local_role();
```

Comment:
This is not a Bluedroid status code AFAICT. [1] So don't test for BT_STATUS_SUCCESS below. I'd also suggest to call the variable |local_role| instead of |status|.

[1] https://git.kernel.org/cgit/bluetooth/bluez.git/tree/android/hal-ipc-api.txt#n705